Src/ILGPU.Algorithms/MatrixOperations/MaskedSparseMatrixExtensions.cs

```diff
@@ -0,0 +1,390 @@
+        /// <param name="outView">
+        /// A dense output view containing the results of the multiplication.
+        /// </param>
+        /// <param name="processor">An instance of an actual mul-add operation.</param>
+        internal static void MaskedSparseMatrixMultiplierKernel<
+            T,
+            TPredicate,
+            TStride,
+            TProcessor>(
+            TPredicate maskPredicate,
+            ArrayView2D<T, TStride> aView,
+            SparseMatrixView<T, TStride> bView,
+            ArrayView2D<T, TStride> outView,
+            TProcessor processor)
+            where T : unmanaged
+            where TStride : struct, IStride2D
+            where TPredicate : struct, InlineList.IPredicate<Index2D>
+            where TProcessor : struct, IMaskedSparseMatrixProcessor<T>
+        {
+            var maskView = SharedMemory.GetDynamic<int>();
+            Trace.Assert(
+                Group.DimX <= maskView.IntLength,
+                "Invalid shared memory config");
+
+            // Get all predicate results and perform thread compaction
+            int index = Grid.GlobalLinearIndex;
+            var maskIndex2D = aView.Stride.ReconstructFromElementIndex(index);
+            bool apply = maskIndex2D.X < aView.Extent.X & maskIndex2D.Y < aView.Extent.Y;
+
+            // Check the predicate (if in range) and compute thread-compaction offsets
+            if (apply)
+                apply = maskPredicate.Apply(maskIndex2D);
+            int groupOffset = GroupExtensions.ExclusiveScan<int, AddInt32>(
+                Utilities.Select(apply, 1, 0));
+
+            // Store current result and determine the total number of matrix entries
+            if (apply)
+                maskView[groupOffset] = index;
+            int numEntries = Group.BarrierPopCount(apply);
+
+            // Load the current entry (if any)
+            if (Group.IdxX >= numEntries)
+                return;
+            int processingIndex = maskView[Group.IdxX];
+            var index2D = aView.Stride.ReconstructFromElementIndex(processingIndex);
+
+            // Load next number of neighbors
+            T dotProduct = default;
+            int numNeighbors = bView.NumNeighbors[index2D.Y];
+
+            // Note that profiling yielded that explicit L1 shared-memory caching did not
+            // help to improve performance in this case. This is also due to the fact
+            // that most reused results automatically end up being fetching into L2 cache
+            // and loaded upon request (based on the structure of the algorithm) into the
+            // L1 parts of the processing unit
+
+            // Implicit transpose of B
+            for (var z = 0; z < numNeighbors; z++)
+            {
+                int columnIndex = bView.Neighbors[index2D.Y, z];
+                dotProduct = processor.MultiplyAdd(
+                    dotProduct,
+                    aView[index2D.X, columnIndex],
```

Comment:
I'm a bit confused by how the masking actually works here. I guess index2D.X will limit the X range we consider. But what about the Y values? That is, if columnIndex takes on a Y value that should be excluded because of the mask, I don't see how this gets excluded from the dotProduct.

Reply:
Currently, unstructured sparsity is expected to affect the entire matrix and thus should also affect the Y coordinates based on thread compaction.

Reply:
I guess I would need to trace it. Basically, I would want to investigate the case where columnIndex takes on a y value that is masked out in A via the predicate. So, I'm not clear what happens when aView is asked to index a value that should be masked out.